opengovsg/plumber
# PLU-125: [EXCEL-5] Connection registration for M365

## Problem
We need to set up the user's plumber folder when the user clicks 'Connect' on our M365 actions

## Solution
Refer to the M365 context doc in our notion for more details, but in general, I make the "connect" button:
1. Create a folder using the user's email in our sharepoint document library, and make it private to us and the user
2. Store the folder ID inside the user's `connection` object.
3. On subsequent loads, check if the user's connection already has a folder ID. If so, we consider the user as successfully connected.

Notes:
- I made `beforeRequest` async so that I can add our M365 auth token to the request. This is needed as the token is protected by a process-wide async mutex.
- We rely on a local `folderId` instead of querying the M365 server to save queries. This means user cannot self-help if they accidentally delete their folder, but I expect this to be rare.

**Before folder creation**
_Showing govtech staging for convenience as I have not created folder there yet_
<img width="500" src="https://github.com/opengovsg/plumber/assets/135598754/f15333c3-3b94-41f9-9f3e-20486f244b48" />

**After folder creation**
<img width="500" src="https://github.com/opengovsg/plumber/assets/135598754/e698189f-8642-431f-a90b-9854ec300068" />

## Tests
- Check that plumber folder is created when user press connect
- Check that users on SharePoint cannot access other users' plumber folder
- Check that our plumber app can access all users' plumber folders
- Regression test: check that other apps with `beforeRequest` did not break, such as telegram.
- Regression test: check FormSG (the only other app with connection registration) registration still works

**Unit tests to be add when team is OK with approach**
